{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#clustering cells from GAE concatenated embeddings and comparison to scRNA clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import torch.nn.functional as F\n",
    "from torch_geometric.nn import GCNConv\n",
    "from torch_geometric.data import Data, DataLoader\n",
    "from sklearn.cluster import KMeans\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import wandb\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import normalize\n",
    "import scanpy as sc\n",
    "from torch.utils.tensorboard import SummaryWriter\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code without tensorboard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Finishing last run (ID:kp08y4wi) before initializing another..."
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "    table.wandb td:nth-child(1) { padding: 0 10px; text-align: left ; width: auto;} td:nth-child(2) {text-align: left ; width: 100%}\n",
       "    .wandb-row { display: flex; flex-direction: row; flex-wrap: wrap; justify-content: flex-start; width: 100% }\n",
       "    .wandb-col { display: flex; flex-direction: column; flex-basis: 100%; flex: 1; padding: 10px; }\n",
       "    </style>\n",
       "<div class=\"wandb-row\"><div class=\"wandb-col\"><h3>Run history:</h3><br/><table class=\"wandb\"><tr><td>loss</td><td>▁█▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁</td></tr><tr><td>val_loss</td><td>█▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁</td></tr></table><br/></div><div class=\"wandb-col\"><h3>Run summary:</h3><br/><table class=\"wandb\"><tr><td>loss</td><td>0.69315</td></tr><tr><td>val_loss</td><td>0.69315</td></tr></table><br/></div></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run <strong style=\"color:#cdcd00\">solar-music-1</strong> at: <a href='https://wandb.ai/atriacaro-university-of-vienna/sc_GAE/runs/kp08y4wi' target=\"_blank\">https://wandb.ai/atriacaro-university-of-vienna/sc_GAE/runs/kp08y4wi</a><br/> View project at: <a href='https://wandb.ai/atriacaro-university-of-vienna/sc_GAE' target=\"_blank\">https://wandb.ai/atriacaro-university-of-vienna/sc_GAE</a><br/>Synced 6 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Find logs at: <code>./wandb/run-20250203_141332-kp08y4wi/logs</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "The new W&B backend becomes opt-out in version 0.18.0; try it out with `wandb.require(\"core\")`! See https://wandb.me/wandb-core for more information."
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Successfully finished last run (ID:kp08y4wi). Initializing new run:<br/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "wandb version 0.19.5 is available!  To upgrade, please run:\n",
       " $ pip install wandb --upgrade"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Tracking run with wandb version 0.17.3"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Run data is saved locally in <code>/Users/work/Library/Mobile Documents/com~apple~CloudDocs/Desktop/ADesktop/Studium/PhD/DataMining/TF_Gene_Interaction_Prediction/src/wandb/run-20250203_143841-t0d3nw1j</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Syncing run <strong><a href='https://wandb.ai/atriacaro-university-of-vienna/sc_GAE/runs/t0d3nw1j' target=\"_blank\">frosty-butterfly-2</a></strong> to <a href='https://wandb.ai/atriacaro-university-of-vienna/sc_GAE' target=\"_blank\">Weights & Biases</a> (<a href='https://wandb.me/run' target=\"_blank\">docs</a>)<br/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View project at <a href='https://wandb.ai/atriacaro-university-of-vienna/sc_GAE' target=\"_blank\">https://wandb.ai/atriacaro-university-of-vienna/sc_GAE</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run at <a href='https://wandb.ai/atriacaro-university-of-vienna/sc_GAE/runs/t0d3nw1j' target=\"_blank\">https://wandb.ai/atriacaro-university-of-vienna/sc_GAE/runs/t0d3nw1j</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "edge_index tensor([[  901,  1482,  1632,  ...,  7939,  8086,  8258],\n",
      "        [    0,     0,     0,  ..., 13133, 13133, 13133]])\n",
      "edge_index shape torch.Size([2, 82068])\n",
      "Validation Loss: 0.6932\n",
      "Epoch 1/100, Loss: 0.6932\n",
      "Validation Loss: 0.6932\n",
      "Epoch 2/100, Loss: 0.6931\n",
      "Validation Loss: 0.6932\n",
      "Epoch 3/100, Loss: 0.6932\n",
      "Validation Loss: 0.6932\n",
      "Epoch 4/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 5/100, Loss: 0.6932\n",
      "Validation Loss: 0.6931\n",
      "Epoch 6/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 7/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 8/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 9/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 10/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 11/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 12/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 13/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 14/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 15/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 16/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 17/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 18/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 19/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 20/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 21/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 22/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 23/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 24/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 25/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 26/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 27/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 28/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 29/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 30/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 31/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 32/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 33/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 34/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 35/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 36/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 37/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 38/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 39/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 40/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 41/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 42/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 43/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 44/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 45/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 46/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 47/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 48/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 49/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 50/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 51/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 52/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 53/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 54/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 55/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 56/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 57/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 58/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 59/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 60/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 61/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 62/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 63/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 64/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 65/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 66/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 67/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 68/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 69/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 70/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 71/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 72/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 73/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 74/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 75/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 76/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 77/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 78/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 79/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 80/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 81/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 82/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 83/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 84/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 85/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 86/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 87/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 88/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 89/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 90/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 91/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 92/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 93/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 94/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 95/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 96/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 97/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 98/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 99/100, Loss: 0.6931\n",
      "Validation Loss: 0.6931\n",
      "Epoch 100/100, Loss: 0.6931\n"
     ]
    }
   ],
   "source": [
    "class GAEModel(torch.nn.Module):\n",
    "    def __init__(self, in_channels, out_channels):\n",
    "        super(GAEModel, self).__init__()\n",
    "        self.conv1 = GCNConv(in_channels, 2 * out_channels)\n",
    "        self.conv2 = GCNConv(2 * out_channels, out_channels)\n",
    "        self.decoder = torch.nn.Bilinear(out_channels, out_channels, 1)\n",
    "\n",
    "    def encode(self, x, edge_index):\n",
    "        x = F.relu(self.conv1(x, edge_index))\n",
    "        return self.conv2(x, edge_index)\n",
    "\n",
    "    def decode(self, z, edge_index):\n",
    "        src, dst = z[edge_index[0]], z[edge_index[1]]\n",
    "        return torch.sigmoid(self.decoder(src, dst))\n",
    "\n",
    "    def forward(self, x, edge_index):\n",
    "        z = self.encode(x, edge_index)\n",
    "        out = self.decode(z, edge_index)\n",
    "        return out\n",
    "    \n",
    "    def get_node_embeddings(self, x, edge_index):\n",
    "        return self.encode(x, edge_index)\n",
    "\n",
    "def load_data(node_feature_path, edge_incidence_path):\n",
    "    # Load gene expression data for each cell\n",
    "    node_features = pd.read_csv(node_feature_path, index_col=0).head(1000)\n",
    "    # Load edge incidence matrix (common to all cells)\n",
    "    edge_incidence = pd.read_csv(edge_incidence_path, index_col=0).T\n",
    "\n",
    "    # Create a node index map based on the gene order in node_features\n",
    "    node_index_map = {gene: idx for idx, gene in enumerate(node_features.columns)}\n",
    "\n",
    "    # Create edge_index from the edge incidence matrix\n",
    "    edge_index = []\n",
    "    for col in edge_incidence.columns:\n",
    "        if col in node_index_map:\n",
    "            target_idx = node_index_map[col]\n",
    "            source_genes = edge_incidence.index[edge_incidence[col] == 1].tolist()\n",
    "            for source in source_genes:\n",
    "                if source in node_index_map:\n",
    "                    source_idx = node_index_map[source]\n",
    "                    edge_index.append([source_idx, target_idx])\n",
    "\n",
    "    edge_index = torch.tensor(edge_index, dtype=torch.long).t().contiguous()\n",
    "    print('edge_index', edge_index)\n",
    "    print('edge_index shape', edge_index.shape)\n",
    "\n",
    "    # Create a list of Data objects for each cell\n",
    "    data_list = []\n",
    "    for _, row in node_features.iterrows():\n",
    "        x = torch.tensor(row.values, dtype=torch.float).unsqueeze(1)\n",
    "        data = Data(x=x, edge_index=edge_index)\n",
    "        data_list.append(data)\n",
    "    return data_list\n",
    "\n",
    "def train(model, dataset, num_epochs=6):\n",
    "    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)\n",
    "    criterion = torch.nn.BCEWithLogitsLoss()\n",
    "    \n",
    "    train_size = int(0.8 * len(dataset))\n",
    "    train_dataset, val_dataset = torch.utils.data.random_split(dataset, [train_size, len(dataset) - train_size])\n",
    "    train_loader = DataLoader(train_dataset, batch_size=1, shuffle=True)\n",
    "    val_loader = DataLoader(val_dataset, batch_size=1, shuffle=False)\n",
    "\n",
    "    for epoch in range(num_epochs):\n",
    "        model.train()\n",
    "        for data in train_loader:\n",
    "            optimizer.zero_grad()\n",
    "            edge_predictions = model(data.x, data.edge_index).squeeze()\n",
    "            loss = criterion(edge_predictions, edge_predictions.detach())\n",
    "            loss.backward()\n",
    "            optimizer.step()\n",
    "            wandb.log({\"loss\": loss.item()})\n",
    "        \n",
    "        evaluate(model, criterion, val_loader)\n",
    "        print(f'Epoch {epoch+1}/{num_epochs}, Loss: {loss.item():.4f}')\n",
    "    return model\n",
    "\n",
    "def evaluate(model, criterion, val_loader):\n",
    "    model.eval()\n",
    "    val_loss = 0\n",
    "    with torch.no_grad():\n",
    "        for data in val_loader:\n",
    "            edge_predictions = model(data.x, data.edge_index).squeeze()\n",
    "            loss = criterion(edge_predictions, edge_predictions.detach())\n",
    "            val_loss += loss.item()\n",
    "    val_loss /= len(val_loader)\n",
    "    print(f'Validation Loss: {val_loss:.4f}')\n",
    "    wandb.log({\"val_loss\": val_loss})\n",
    "\n",
    "def extract_cell_embeddings(model, dataset):\n",
    "    model.eval()\n",
    "    cell_embeddings = []\n",
    "    with torch.no_grad():\n",
    "        for data in dataset:\n",
    "            z = model.get_node_embeddings(data.x, data.edge_index)\n",
    "            cell_embedding = z.flatten().cpu().numpy()\n",
    "            cell_embeddings.append(cell_embedding)\n",
    "    return np.array(cell_embeddings)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    wandb.init(project=\"sc_GAE\")\n",
    "    \n",
    "    # Load gene expression and interaction data\n",
    "    data_list = load_data('/Users/work/Desktop/expression_matrix_yitao.csv', '/Users/work/Desktop/suberites_presence_absence_yitao.csv')\n",
    "\n",
    "    model = GAEModel(in_channels=1, out_channels=8)\n",
    "    trained_model = train(model, data_list, num_epochs=100)\n",
    "    \n",
    "    cell_embeddings = extract_cell_embeddings(trained_model, data_list)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "edge_index tensor([[  901,  1482,  1632,  ...,  7939,  8086,  8258],\n",
      "        [    0,     0,     0,  ..., 13133, 13133, 13133]])\n",
      "edge_index shape torch.Size([2, 82068])\n"
     ]
    }
   ],
   "source": [
    "data_list = load_data('/Users/work/Desktop/expression_matrix_yitao.csv', '/Users/work/Desktop/suberites_presence_absence_yitao.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>SUB2.g1</th>\n",
       "      <th>SUB2.g2</th>\n",
       "      <th>SUB2.g3</th>\n",
       "      <th>SUB2.g4</th>\n",
       "      <th>SUB2.g5</th>\n",
       "      <th>SUB2.g6</th>\n",
       "      <th>SUB2.g7</th>\n",
       "      <th>SUB2.g8</th>\n",
       "      <th>SUB2.g9</th>\n",
       "      <th>...</th>\n",
       "      <th>SUB2.g13135</th>\n",
       "      <th>SUB2.g13136</th>\n",
       "      <th>SUB2.g13137</th>\n",
       "      <th>SUB2.g13138</th>\n",
       "      <th>SUB2.g13139</th>\n",
       "      <th>SUB2.g13140</th>\n",
       "      <th>SUB2.g13141</th>\n",
       "      <th>SUB2.g13142</th>\n",
       "      <th>SUB2.g13143</th>\n",
       "      <th>SUB2.g13144</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGGACAGCT-1</td>\n",
       "      <td>0.009687</td>\n",
       "      <td>-1.032545</td>\n",
       "      <td>-1.566363</td>\n",
       "      <td>1.018223</td>\n",
       "      <td>-0.933981</td>\n",
       "      <td>-0.410398</td>\n",
       "      <td>-0.349553</td>\n",
       "      <td>-0.444472</td>\n",
       "      <td>-0.132447</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.293056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.365098</td>\n",
       "      <td>-0.652811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGGGTTTCT-1</td>\n",
       "      <td>1.598354</td>\n",
       "      <td>-0.300789</td>\n",
       "      <td>-0.202161</td>\n",
       "      <td>-0.002592</td>\n",
       "      <td>0.181964</td>\n",
       "      <td>0.269252</td>\n",
       "      <td>-0.393041</td>\n",
       "      <td>0.037609</td>\n",
       "      <td>0.115123</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.192047</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.296943</td>\n",
       "      <td>-0.677305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCACAAATGGTA-1</td>\n",
       "      <td>-0.777102</td>\n",
       "      <td>-0.792809</td>\n",
       "      <td>-1.281675</td>\n",
       "      <td>1.262175</td>\n",
       "      <td>-0.185155</td>\n",
       "      <td>-0.726397</td>\n",
       "      <td>-0.455605</td>\n",
       "      <td>-0.671795</td>\n",
       "      <td>-0.448440</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.315716</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.205751</td>\n",
       "      <td>0.327006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCATCGAGAAGC-1</td>\n",
       "      <td>0.696619</td>\n",
       "      <td>-0.306533</td>\n",
       "      <td>0.700373</td>\n",
       "      <td>-1.587061</td>\n",
       "      <td>-0.054666</td>\n",
       "      <td>0.263420</td>\n",
       "      <td>-0.105024</td>\n",
       "      <td>0.250046</td>\n",
       "      <td>0.661372</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.300775</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.310817</td>\n",
       "      <td>-0.878208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCATCTTGTTAC-1</td>\n",
       "      <td>-1.319795</td>\n",
       "      <td>-1.191836</td>\n",
       "      <td>-1.658408</td>\n",
       "      <td>1.626106</td>\n",
       "      <td>-0.142163</td>\n",
       "      <td>-1.176196</td>\n",
       "      <td>-0.391015</td>\n",
       "      <td>-0.537126</td>\n",
       "      <td>-0.254368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.301799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.303617</td>\n",
       "      <td>-0.420889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2666</th>\n",
       "      <td>TTTGGTTGTGATGAAT-1</td>\n",
       "      <td>1.373431</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.041506</td>\n",
       "      <td>-0.439186</td>\n",
       "      <td>0.454166</td>\n",
       "      <td>0.649058</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.291616</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.320812</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.169915</td>\n",
       "      <td>0.400232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2667</th>\n",
       "      <td>TTTGGTTGTTATAGCC-1</td>\n",
       "      <td>1.079402</td>\n",
       "      <td>0.559011</td>\n",
       "      <td>-0.162774</td>\n",
       "      <td>-0.475533</td>\n",
       "      <td>-0.359524</td>\n",
       "      <td>0.943268</td>\n",
       "      <td>-0.280327</td>\n",
       "      <td>-0.424414</td>\n",
       "      <td>-0.203679</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.842400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.301879</td>\n",
       "      <td>0.520434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2668</th>\n",
       "      <td>TTTGTTGAGAGAGTGA-1</td>\n",
       "      <td>-1.486299</td>\n",
       "      <td>-1.097147</td>\n",
       "      <td>-1.712721</td>\n",
       "      <td>1.434417</td>\n",
       "      <td>0.358839</td>\n",
       "      <td>-1.098839</td>\n",
       "      <td>-0.345341</td>\n",
       "      <td>-0.442162</td>\n",
       "      <td>-0.121341</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.292260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.370699</td>\n",
       "      <td>-0.793447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2669</th>\n",
       "      <td>TTTGTTGAGAGGTCGT-1</td>\n",
       "      <td>1.075499</td>\n",
       "      <td>0.579794</td>\n",
       "      <td>1.183538</td>\n",
       "      <td>-0.750108</td>\n",
       "      <td>0.465700</td>\n",
       "      <td>1.618033</td>\n",
       "      <td>-0.163200</td>\n",
       "      <td>-0.444858</td>\n",
       "      <td>-0.234243</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.237166</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.313765</td>\n",
       "      <td>0.076373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2670</th>\n",
       "      <td>TTTGTTGAGGTAACTA-1</td>\n",
       "      <td>1.190661</td>\n",
       "      <td>-0.675838</td>\n",
       "      <td>1.015256</td>\n",
       "      <td>0.327037</td>\n",
       "      <td>-0.050720</td>\n",
       "      <td>1.192914</td>\n",
       "      <td>-0.462873</td>\n",
       "      <td>-0.610539</td>\n",
       "      <td>-0.481627</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.318213</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.027347</td>\n",
       "      <td>0.408016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2671 rows × 13145 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Unnamed: 0   SUB2.g1   SUB2.g2   SUB2.g3   SUB2.g4   SUB2.g5  \\\n",
       "0     AAACCCAAGGACAGCT-1  0.009687 -1.032545 -1.566363  1.018223 -0.933981   \n",
       "1     AAACCCAAGGGTTTCT-1  1.598354 -0.300789 -0.202161 -0.002592  0.181964   \n",
       "2     AAACCCACAAATGGTA-1 -0.777102 -0.792809 -1.281675  1.262175 -0.185155   \n",
       "3     AAACCCATCGAGAAGC-1  0.696619 -0.306533  0.700373 -1.587061 -0.054666   \n",
       "4     AAACCCATCTTGTTAC-1 -1.319795 -1.191836 -1.658408  1.626106 -0.142163   \n",
       "...                  ...       ...       ...       ...       ...       ...   \n",
       "2666  TTTGGTTGTGATGAAT-1  1.373431  2.000000  1.041506 -0.439186  0.454166   \n",
       "2667  TTTGGTTGTTATAGCC-1  1.079402  0.559011 -0.162774 -0.475533 -0.359524   \n",
       "2668  TTTGTTGAGAGAGTGA-1 -1.486299 -1.097147 -1.712721  1.434417  0.358839   \n",
       "2669  TTTGTTGAGAGGTCGT-1  1.075499  0.579794  1.183538 -0.750108  0.465700   \n",
       "2670  TTTGTTGAGGTAACTA-1  1.190661 -0.675838  1.015256  0.327037 -0.050720   \n",
       "\n",
       "       SUB2.g6   SUB2.g7   SUB2.g8   SUB2.g9  ...  SUB2.g13135  SUB2.g13136  \\\n",
       "0    -0.410398 -0.349553 -0.444472 -0.132447  ...          0.0          0.0   \n",
       "1     0.269252 -0.393041  0.037609  0.115123  ...          0.0          0.0   \n",
       "2    -0.726397 -0.455605 -0.671795 -0.448440  ...          0.0          0.0   \n",
       "3     0.263420 -0.105024  0.250046  0.661372  ...          0.0          0.0   \n",
       "4    -1.176196 -0.391015 -0.537126 -0.254368  ...          0.0          0.0   \n",
       "...        ...       ...       ...       ...  ...          ...          ...   \n",
       "2666  0.649058  2.000000  2.000000  0.291616  ...          0.0          0.0   \n",
       "2667  0.943268 -0.280327 -0.424414 -0.203679  ...          0.0          0.0   \n",
       "2668 -1.098839 -0.345341 -0.442162 -0.121341  ...          0.0          0.0   \n",
       "2669  1.618033 -0.163200 -0.444858 -0.234243  ...          0.0          0.0   \n",
       "2670  1.192914 -0.462873 -0.610539 -0.481627  ...          0.0          0.0   \n",
       "\n",
       "      SUB2.g13137  SUB2.g13138  SUB2.g13139  SUB2.g13140  SUB2.g13141  \\\n",
       "0             0.0    -0.293056          0.0          0.0          0.0   \n",
       "1             0.0    -0.192047          0.0          0.0          0.0   \n",
       "2             0.0    -0.315716          0.0          0.0          0.0   \n",
       "3             0.0    -0.300775          0.0          0.0          0.0   \n",
       "4             0.0    -0.301799          0.0          0.0          0.0   \n",
       "...           ...          ...          ...          ...          ...   \n",
       "2666          0.0    -0.320812          0.0          0.0          0.0   \n",
       "2667          0.0     0.842400          0.0          0.0          0.0   \n",
       "2668          0.0    -0.292260          0.0          0.0          0.0   \n",
       "2669          0.0    -0.237166          0.0          0.0          0.0   \n",
       "2670          0.0    -0.318213          0.0          0.0          0.0   \n",
       "\n",
       "      SUB2.g13142  SUB2.g13143  SUB2.g13144  \n",
       "0             0.0    -0.365098    -0.652811  \n",
       "1             0.0    -0.296943    -0.677305  \n",
       "2             0.0    -0.205751     0.327006  \n",
       "3             0.0    -0.310817    -0.878208  \n",
       "4             0.0    -0.303617    -0.420889  \n",
       "...           ...          ...          ...  \n",
       "2666          0.0    -0.169915     0.400232  \n",
       "2667          0.0    -0.301879     0.520434  \n",
       "2668          0.0    -0.370699    -0.793447  \n",
       "2669          0.0    -0.313765     0.076373  \n",
       "2670          0.0    -0.027347     0.408016  \n",
       "\n",
       "[2671 rows x 13145 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"/Users/work/Desktop/expression_matrix_yitao.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/tf_gene_interaction/lib/python3.12/site-packages/sklearn/manifold/_spectral_embedding.py:329: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from sklearn.cluster import SpectralClustering\n",
    "\n",
    "\n",
    "def cluster_cells(cell_embeddings, n_clusters=2):\n",
    "    spectral = SpectralClustering(n_clusters=n_clusters, affinity='nearest_neighbors', random_state=0)\n",
    "    return spectral.fit_predict(cell_embeddings)  # Returns cluster labels for each cell\n",
    "\n",
    "\n",
    "# def cluster_cells(cell_embeddings, n_clusters=2):\n",
    "#     kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(cell_embeddings)\n",
    "#     return kmeans.labels_  # Returns cluster labels for each cell\n",
    "\n",
    "cluster_labels = cluster_cells(cell_embeddings, n_clusters=15)\n",
    "\n",
    "   \n",
    "#    Save clustering results\n",
    "pd.DataFrame(cluster_labels, columns=['Cluster']).to_csv(\"cell_clusters.csv\", index=False)\n",
    "#cluster_labels = pd.read_csv(\"/Users/work/Library/Mobile Documents/com~apple~CloudDocs/Desktop/ADesktop/Studium/PhD/DataMining/TF_Gene_Interaction_Prediction/src/cell_clusters.csv\")[\"Cluster\"]\n",
    "import scanpy as sc\n",
    "adata = sc.read_h5ad(\"/Users/work/Desktop/subdom_processed.h5ad\")\n",
    "adata.obs_names = np.arange(0, len(adata)).astype(str)\n",
    "labels_adata = adata.obs['clusters']\n",
    "\n",
    "from sklearn.metrics import adjusted_rand_score\n",
    "\n",
    "# Assuming `labels_ml` contains the cluster labels from your ML pipeline\n",
    "# and that `labels_ml` and `labels_adata` are aligned and of the same length.\n",
    "\n",
    "# Calculate Adjusted Rand Index\n",
    "ari = adjusted_rand_score(labels_adata, cluster_labels)\n",
    "\n",
    "# Print the ARI result\n",
    "print(f\"Adjusted Rand Index (ARI) spectral: {ari}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparison to un-embedded data #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"/Users/work/Desktop/subdom_processed.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_matrix =adata.raw.X.toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_matrix =adata.X\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "k = 15  # Specify the number of clusters\n",
    "kmeans = KMeans(n_clusters=k, random_state=42)\n",
    "adata.obs['kmeans_clusters'] = kmeans.fit_predict(data_matrix)\n",
    "kmeans_pre_embedding = adata.obs['kmeans_clusters']\n",
    "\n",
    "kmeans = KMeans(n_clusters=k, random_state=42, init='k-means++', n_init=10)\n",
    "spherical_pre_embedding = kmeans.fit_predict(data_matrix)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adjusted Rand Index (ARI): (0.5623482165003406, 0.6224136833939046)\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import adjusted_rand_score\n",
    "\n",
    "# Assuming `labels_ml` contains the cluster labels from your ML pipeline\n",
    "# and that `labels_ml` and `labels_adata` are aligned and of the same length.\n",
    "\n",
    "# Calculate Adjusted Rand Index\n",
    "ari = adjusted_rand_score(labels_adata, kmeans_pre_embedding)\n",
    "ari2 = adjusted_rand_score(labels_adata, spherical_pre_embedding)\n",
    "\n",
    "\n",
    "# Print the ARI result\n",
    "print(f\"Adjusted Rand Index (ARI): {ari,ari2}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# time line analysis #\n",
    "\n",
    "Average gene expression over clusters (and later metacells) to construct timeline interactions of TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"/Users/work/Desktop/subdom_processed.h5ad\")\n",
    "adata.obs_names = np.arange(0, len(adata)).astype(str)\n",
    "labels_adata = adata.obs['clusters']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>ct_gene_corr</th>\n",
       "      <th>ct_correlates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SUB2.g1</th>\n",
       "      <td>SUB2.g1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>219</td>\n",
       "      <td>0.089854</td>\n",
       "      <td>91.800824</td>\n",
       "      <td>240.0</td>\n",
       "      <td>False</td>\n",
       "      <td>4.980643e-02</td>\n",
       "      <td>-4.542987</td>\n",
       "      <td>-0.922765</td>\n",
       "      <td>7.669924e-12</td>\n",
       "      <td>0.022066</td>\n",
       "      <td>0.668407</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g2</th>\n",
       "      <td>SUB2.g2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>364</td>\n",
       "      <td>0.163984</td>\n",
       "      <td>86.372145</td>\n",
       "      <td>438.0</td>\n",
       "      <td>True</td>\n",
       "      <td>8.974101e-02</td>\n",
       "      <td>-2.857633</td>\n",
       "      <td>0.520586</td>\n",
       "      <td>-2.021063e-11</td>\n",
       "      <td>0.062651</td>\n",
       "      <td>0.507098</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g3</th>\n",
       "      <td>SUB2.g3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>592</td>\n",
       "      <td>0.269188</td>\n",
       "      <td>77.836016</td>\n",
       "      <td>719.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.482865e-01</td>\n",
       "      <td>-3.145663</td>\n",
       "      <td>0.273915</td>\n",
       "      <td>-1.628408e-11</td>\n",
       "      <td>0.071940</td>\n",
       "      <td>0.673111</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g4</th>\n",
       "      <td>SUB2.g4</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>258</td>\n",
       "      <td>0.107450</td>\n",
       "      <td>90.340696</td>\n",
       "      <td>287.0</td>\n",
       "      <td>False</td>\n",
       "      <td>6.187559e-02</td>\n",
       "      <td>-4.124707</td>\n",
       "      <td>-0.564547</td>\n",
       "      <td>-7.620040e-12</td>\n",
       "      <td>0.029855</td>\n",
       "      <td>-0.288837</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g5</th>\n",
       "      <td>SUB2.g5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>107</td>\n",
       "      <td>0.043804</td>\n",
       "      <td>95.994010</td>\n",
       "      <td>117.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2.365992e-02</td>\n",
       "      <td>-4.478562</td>\n",
       "      <td>-0.867591</td>\n",
       "      <td>-1.315720e-12</td>\n",
       "      <td>0.015843</td>\n",
       "      <td>0.379239</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g13140</th>\n",
       "      <td>SUB2.g13140</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g13141</th>\n",
       "      <td>SUB2.g13141</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g13142</th>\n",
       "      <td>SUB2.g13142</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g13143</th>\n",
       "      <td>SUB2.g13143</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000749</td>\n",
       "      <td>99.925122</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>5.678186e-04</td>\n",
       "      <td>-4.993979</td>\n",
       "      <td>-1.308999</td>\n",
       "      <td>1.392674e-13</td>\n",
       "      <td>0.001943</td>\n",
       "      <td>0.188750</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUB2.g13144</th>\n",
       "      <td>SUB2.g13144</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>isoseq_reference</td>\n",
       "      <td>205</td>\n",
       "      <td>0.080494</td>\n",
       "      <td>92.324972</td>\n",
       "      <td>215.0</td>\n",
       "      <td>False</td>\n",
       "      <td>4.380609e-02</td>\n",
       "      <td>-4.413703</td>\n",
       "      <td>-0.812045</td>\n",
       "      <td>7.782973e-12</td>\n",
       "      <td>0.021736</td>\n",
       "      <td>-0.039962</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13144 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_ids    feature_types            genome  \\\n",
       "SUB2.g1          SUB2.g1  Gene Expression  isoseq_reference   \n",
       "SUB2.g2          SUB2.g2  Gene Expression  isoseq_reference   \n",
       "SUB2.g3          SUB2.g3  Gene Expression  isoseq_reference   \n",
       "SUB2.g4          SUB2.g4  Gene Expression  isoseq_reference   \n",
       "SUB2.g5          SUB2.g5  Gene Expression  isoseq_reference   \n",
       "...                  ...              ...               ...   \n",
       "SUB2.g13140  SUB2.g13140  Gene Expression  isoseq_reference   \n",
       "SUB2.g13141  SUB2.g13141  Gene Expression  isoseq_reference   \n",
       "SUB2.g13142  SUB2.g13142  Gene Expression  isoseq_reference   \n",
       "SUB2.g13143  SUB2.g13143  Gene Expression  isoseq_reference   \n",
       "SUB2.g13144  SUB2.g13144  Gene Expression  isoseq_reference   \n",
       "\n",
       "             n_cells_by_counts  mean_counts  pct_dropout_by_counts  \\\n",
       "SUB2.g1                    219     0.089854              91.800824   \n",
       "SUB2.g2                    364     0.163984              86.372145   \n",
       "SUB2.g3                    592     0.269188              77.836016   \n",
       "SUB2.g4                    258     0.107450              90.340696   \n",
       "SUB2.g5                    107     0.043804              95.994010   \n",
       "...                        ...          ...                    ...   \n",
       "SUB2.g13140                  0     0.000000             100.000000   \n",
       "SUB2.g13141                  0     0.000000             100.000000   \n",
       "SUB2.g13142                  0     0.000000             100.000000   \n",
       "SUB2.g13143                  2     0.000749              99.925122   \n",
       "SUB2.g13144                205     0.080494              92.324972   \n",
       "\n",
       "             total_counts  highly_variable         means  dispersions  \\\n",
       "SUB2.g1             240.0            False  4.980643e-02    -4.542987   \n",
       "SUB2.g2             438.0             True  8.974101e-02    -2.857633   \n",
       "SUB2.g3             719.0            False  1.482865e-01    -3.145663   \n",
       "SUB2.g4             287.0            False  6.187559e-02    -4.124707   \n",
       "SUB2.g5             117.0            False  2.365992e-02    -4.478562   \n",
       "...                   ...              ...           ...          ...   \n",
       "SUB2.g13140           0.0            False  1.000000e-12          NaN   \n",
       "SUB2.g13141           0.0            False  1.000000e-12          NaN   \n",
       "SUB2.g13142           0.0            False  1.000000e-12          NaN   \n",
       "SUB2.g13143           2.0            False  5.678186e-04    -4.993979   \n",
       "SUB2.g13144         215.0            False  4.380609e-02    -4.413703   \n",
       "\n",
       "             dispersions_norm          mean       std  ct_gene_corr  \\\n",
       "SUB2.g1             -0.922765  7.669924e-12  0.022066      0.668407   \n",
       "SUB2.g2              0.520586 -2.021063e-11  0.062651      0.507098   \n",
       "SUB2.g3              0.273915 -1.628408e-11  0.071940      0.673111   \n",
       "SUB2.g4             -0.564547 -7.620040e-12  0.029855     -0.288837   \n",
       "SUB2.g5             -0.867591 -1.315720e-12  0.015843      0.379239   \n",
       "...                       ...           ...       ...           ...   \n",
       "SUB2.g13140          0.000000  0.000000e+00  1.000000           NaN   \n",
       "SUB2.g13141          0.000000  0.000000e+00  1.000000           NaN   \n",
       "SUB2.g13142          0.000000  0.000000e+00  1.000000           NaN   \n",
       "SUB2.g13143         -1.308999  1.392674e-13  0.001943      0.188750   \n",
       "SUB2.g13144         -0.812045  7.782973e-12  0.021736     -0.039962   \n",
       "\n",
       "             ct_correlates  \n",
       "SUB2.g1              False  \n",
       "SUB2.g2              False  \n",
       "SUB2.g3              False  \n",
       "SUB2.g4              False  \n",
       "SUB2.g5              False  \n",
       "...                    ...  \n",
       "SUB2.g13140          False  \n",
       "SUB2.g13141          False  \n",
       "SUB2.g13142          False  \n",
       "SUB2.g13143          False  \n",
       "SUB2.g13144          False  \n",
       "\n",
       "[13144 rows x 15 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/tb/mtqjrwzn04v2ws67ymnhmv5h0000gp/T/ipykernel_1524/623520819.py:18: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  cluster_avg_expression = expression_df.groupby('cluster').mean()\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Assuming `adata` is your AnnData object and cluster labels are in `adata.obs['clusters']`\n",
    "clusters = adata.obs['clusters']\n",
    "\n",
    "# Convert the expression matrix to a DataFrame for easy manipulation\n",
    "expression_df = pd.DataFrame(\n",
    "    adata.X.toarray() if hasattr(adata.X, \"toarray\") else adata.X, \n",
    "    index=adata.obs_names, \n",
    "    columns=adata.var_names\n",
    ")\n",
    "\n",
    "# Add cluster information to the expression DataFrame\n",
    "expression_df['cluster'] = clusters.values\n",
    "\n",
    "# Group by cluster and calculate the mean for each gene\n",
    "cluster_avg_expression = expression_df.groupby('cluster').mean()\n",
    "\n",
    "# Drop the cluster column after grouping\n",
    "cluster_avg_expression = cluster_avg_expression.drop(columns='cluster', errors='ignore')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cluster_avg_expression.to_csv(\"cluster_avg_expression.csv\")\n",
    "cluster_avg_expression = pd.read_csv(\"/Users/work/Library/Mobile Documents/com~apple~CloudDocs/Desktop/ADesktop/Studium/PhD/DataMining/TF_Gene_Interaction_Prediction/src/cluster_avg_expression.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "two cluster approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   outgoing_gene incoming_gene  outgoing_gene_expr  incoming_gene_expr\n",
      "3        SUB2.g1     SUB2.g902            0.466621            0.031868\n",
      "11       SUB2.g1    SUB2.g1483            0.466621           -0.521559\n",
      "13       SUB2.g1    SUB2.g1633            0.466621           -0.435638\n",
      "14       SUB2.g1    SUB2.g1666            0.466621           -0.281713\n",
      "16       SUB2.g1    SUB2.g1976            0.466621            0.290553\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load your data (replace with actual file paths)\n",
    "gene_expression = cluster_avg_expression.T\n",
    "adj_matrix = pd.read_csv(\"/Users/work/Desktop/suberites_presence_absence_yitao.csv\",index_col=0)\n",
    "# Select only the relevant cell types (0 and 7)\n",
    "expression_filtered = gene_expression[[0, 7]]\n",
    "# Convert adjacency matrix to a long format (edges list)\n",
    "edges = adj_matrix.stack().reset_index()\n",
    "edges.columns = [\"outgoing_gene\", \"incoming_gene\", \"interaction\"]\n",
    "edges = edges[edges[\"interaction\"] == 1]  # Keep only existing interactions\n",
    "\n",
    "# Merge to get expression values\n",
    "edges = edges.merge(expression_filtered[0], left_on=\"outgoing_gene\", right_index=True)\n",
    "edges = edges.merge(expression_filtered[7], left_on=\"incoming_gene\", right_index=True)\n",
    "\n",
    "# Rename columns for clarity\n",
    "edges = edges.rename(columns={0: \"outgoing_gene_expr\", 7: \"incoming_gene_expr\"})\n",
    "\n",
    "del edges[\"interaction\"]\n",
    "# Save the final table\n",
    "edges.to_csv(\"filtered_gene_interactions.csv\", index=False)\n",
    "\n",
    "print(edges.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['outgoing_gene', 'incoming_gene', 'outgoing_gene_expr',\n",
       "       'incoming_gene_expr'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## conditional probability ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# take all cells of both of the clusters, train MLP on these, cluster the conditional probabilities;\n",
    "# P(cell5 | gene 1)\n",
    "\n",
    "\n",
    "#first take all the cells for the separate clusters, only using the genes from cluster 0 that appear in the adjacency matrix\n",
    "#the caculate the conditional probailites and train the MLP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor([[0.3707, 0.3358, 0.2935],\n",
      "        [0.3051, 0.3007, 0.3942],\n",
      "        [0.2773, 0.2813, 0.4414],\n",
      "        [0.4571, 0.3270, 0.2159],\n",
      "        [0.3077, 0.3373, 0.3550],\n",
      "        [0.3444, 0.3422, 0.3134],\n",
      "        [0.3614, 0.3331, 0.3055],\n",
      "        [0.3348, 0.3431, 0.3222],\n",
      "        [0.2486, 0.2454, 0.5060],\n",
      "        [0.3400, 0.3414, 0.3186]], grad_fn=<SoftmaxBackward0>)\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "\n",
    "# Define an MLP for conditional probability estimation\n",
    "class ConditionalMLP(nn.Module):\n",
    "    def __init__(self, input_dim, hidden_dim, output_dim):\n",
    "        super(ConditionalMLP, self).__init__()\n",
    "        self.network = nn.Sequential(\n",
    "            nn.Linear(input_dim, hidden_dim),\n",
    "            nn.ReLU(),\n",
    "            nn.Linear(hidden_dim, output_dim),\n",
    "            nn.Softmax(dim=-1)  # Outputs conditional probabilities\n",
    "        )\n",
    "\n",
    "    def forward(self, x):\n",
    "        return self.network(x)\n",
    "\n",
    "# Example usage\n",
    "model = ConditionalMLP(input_dim=2, hidden_dim=32, output_dim=3)  # 3 output clusters\n",
    "x = torch.randn(10, 2)  # 10 samples with 2 features each\n",
    "probs = model(x)\n",
    "print(probs)  # Outputs probability distribution over 3 clusters\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
